File: live_project/flutter_live_app/lib/features/live/data/models/live_room.dart
/// 客户端表示一个直播间的模型。
///
/// fromJson 兼容后端 camelCase 和数据库常见的 snake_case，并为缺失字段提供
/// 安全默认值，避免不完整数据直接导致页面崩溃。
class LiveRoom {
  const LiveRoom({
    required this.id,
    required this.title,
    required this.anchorName,
    required this.anchorAvatar,
    required this.onlineCount,
    required this.coverUrl,
    required this.status,
    required this.playUrl,
    required this.pushUrl,
    required this.streamName,
    required this.category,
    this.createdAt,
    this.updatedAt,
  });

  final int id;
  final String title;
  final String anchorName;
  final String anchorAvatar;
  final int onlineCount;
  final String coverUrl;
  final String status;
  final String playUrl;
  final String pushUrl;
  final String streamName;
  final String category;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  factory LiveRoom.fromJson(Map<String, Object?> json) {
    // 所有服务端字段在这里集中转换，页面只使用类型明确的 Dart 字段。
    return LiveRoom(
      id: _int(json['id']),
      title: _string(json['title']),
      anchorName: _string(json['anchorName'] ?? json['anchor_name']),
      anchorAvatar: _string(json['anchorAvatar'] ?? json['anchor_avatar']),
      onlineCount: _int(json['onlineCount'] ?? json['online_count']),
      coverUrl: _string(json['coverUrl'] ?? json['cover_url']),
      status: _string(json['status'], fallback: 'living'),
      playUrl: _string(json['playUrl'] ?? json['play_url']),
      pushUrl: _string(json['pushUrl'] ?? json['push_url']),
      streamName: _string(json['streamName'] ?? json['stream_name']),
      category: _string(json['category']),
      createdAt: _date(json['createdAt'] ?? json['created_at']),
      updatedAt: _date(json['updatedAt'] ?? json['updated_at']),
    );
  }

  Map<String, Object?> toJson() {
    // toJson 方便缓存、日志和后续向其他接口传递房间快照。
    return <String, Object?>{
      'id': id,
      'title': title,
      'anchorName': anchorName,
      'anchorAvatar': anchorAvatar,
      'onlineCount': onlineCount,
      'coverUrl': coverUrl,
      'status': status,
      'playUrl': playUrl,
      'pushUrl': pushUrl,
      'streamName': streamName,
      'category': category,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  static String _string(Object? value, {String fallback = ''}) =>
      value is String && value.isNotEmpty ? value : fallback;

  static int _int(Object? value) =>
      value is int ? value : int.tryParse('$value') ?? 0;

  static DateTime? _date(Object? value) =>
      DateTime.tryParse(value as String? ?? '');
}

File: live_project/flutter_live_server/app/schemas/live_room.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class CreateLiveRoomRequest(BaseModel):
    """主播点击“开始直播”前提交的房间信息。

    开播接口要求登录；anchor_name 仍保留用于兼容旧客户端，但服务端会用当前
    登录用户的 display_name 覆盖主播身份，不能由客户端伪造。
    """

    model_config = ConfigDict(populate_by_name=True)

    title: str = Field(min_length=1, max_length=255)
    anchor_name: str = Field(
        min_length=1,
        max_length=100,
        alias="anchorName",
    )
    category: str = Field(default="综合", max_length=100)


class LiveRoomResponse(BaseModel):
    """数据库 LiveRoom 对外暴露的安全响应格式。

    serialization_alias 把 Python/数据库字段转换成 Flutter 更习惯的 camelCase。
    """

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: int
    title: str
    anchor_name: str = Field(serialization_alias="anchorName")
    anchor_avatar: str = Field(serialization_alias="anchorAvatar")
    online_count: int = Field(serialization_alias="onlineCount")
    cover_url: str = Field(serialization_alias="coverUrl")
    status: str
    play_url: str = Field(serialization_alias="playUrl")
    category: str
    following: bool = False
    liked: bool = False
    like_count: int = Field(default=0, serialization_alias="likeCount")
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class LiveRoomHostResponse(LiveRoomResponse):
    """只给主播控制面返回的房间响应。

    pushUrl 是写入 SRS 的凭证式地址，不能出现在观众列表和详情接口中。
    """

    push_url: str = Field(serialization_alias="pushUrl")
    stream_name: str = Field(serialization_alias="streamName")

    @field_validator("push_url", "stream_name", mode="before")
    @classmethod
    def empty_stream_fields_are_safe(cls, value: object) -> str:
        """兼容迁移前的旧 ORM 对象或历史数据中的 NULL。"""
        return value if isinstance(value, str) else ""
